### PR TITLE
[IMP] tools: remove duplicate occurrences from PO files

### PR DESCRIPTION
**Pull request by oomsveta**
Code translation references have all their line numbers set to 0 in order to avoid unnecessary diff when moving code around. This leads to situations where two references are exactly the same, adding more lines for no benefit.

This commit adapts the translation export code to properly remove all duplicate entries.

Task-3948298

### CODE REVIEW

File: odoo/tools/translate.py [hunk 1]
```diff
@@ -842,18 +842,19 @@ def add_entry(self, modules, tnrs, source, trad, comments=None):
...
-        for typy, name, res_id in tnrs:
-            if typy == 'code':
-                code = True
-                res_id = 0
-            if isinstance(res_id, int) or res_id.isdigit():
-                # second term of occurrence must be a digit
-                # occurrence line at 0 are discarded when rendered to string
-                entry.occurrences.append((u"%s:%s" % (typy, name), str(res_id)))
+        occurrences = set()
+        for type_, *ref in tnrs:
```
Comment (MissingNoShiny) [inline odoo/tools/translate.py:846]: I would unpack everything here, it would avoid unpacking `ref` later and the awkward `join`

**Review by HydrionBurst — APPROVED** [review]
LGTM

File: odoo/tools/translate.py [hunk 2]
```diff
@@ -842,18 +842,19 @@ def add_entry(self, modules, tnrs, source, trad, comments=None):
...
+                name = f"code:{fpath}"
+                # lineno is set to 0 to avoid creating diff in PO files every
+                # time the code is moved around
+                lineno = "0"
             else:
-                entry.occurrences.append((u"%s:%s:%s" % (typy, name, res_id), ''))
+                name = f"{type_}:" + ":".join(ref)
+                lineno = None   # no lineno for model/model_terms sources
+            occurrences.add((name, lineno))
+        entry.occurrences = sorted(occurrences)
```
[the remark this answers is not in the document]
Comment (oomsveta) [reply odoo/tools/translate.py:857]: The thing is, sets are not ordered in Python. But you're right, it's better not to sort it there. A workaround could be to store everything as a list and deduplicate with `list(dict.fromkeys(occurrences))`. What do you think?

**Comment by MissingNoShiny** [comment]
robodoo r+

**Pull request closed without merging**